### PR TITLE
make polkadot-runtime optional feature

### PR DESCRIPTION
**Pull request by ordian**
See https://github.com/paritytech/parity-bridges-common/pull/1053#discussion_r700312155 for the context.

### CODE REVIEW

**Review by tomusdrw — APPROVED** [review]
Great, thanks!

If we want to avoid accidentally disabling all of the runtimes, perhaps introducing another feature `no-runtimes` and throwing a compilation error in case it's not explicitly enabled is reasonable?

**Review by bkchr — CHANGES_REQUESTED** [review]
The code isn't adapted for these feature changes.

**Comment by ordian** [comment]
> The code isn't adapted for these feature changes.

could you elaborate? do you mean
> introducing another feature no-runtimes and throwing a compilation error in case it's not explicitly enabled

**Comment by bkchr** [comment]
I mean, if you disable the `polkadot` feature for the `client` or `service`, then both would not compile. But @tomusdrw requires the generic functions of `polkadot-service` for example to run it with its own runtime etc.

**Comment by bkchr** [comment]
Ahh I see that all usage is hidden behind the `FullNode` feature. However I would still favor that https://github.com/paritytech/polkadot/blob/master/node/service/src/lib.rs#L1356-L1366 this is feature gated and all the other usages of `polkadot-runtime`. Like it is done for the other runtimes.

Same applied for the client crate.

We can just throw a runtime error if someone calls the functions and no runtime is enabled.

**Review by bkchr — APPROVED** [review]

**Review by tomusdrw — COMMENTED** [review]
Shall we add a CI check to make sure the `no-default-features` compilation works?

**Comment by bkchr** [comment]
Fine by me.

**Comment by ordian** [comment]
> Shall we add a CI check to make sure the no-default-features compilation works?

It doesn't work for `polkadot-client` and you will need at least one runtime when compiling `cli` and `service` with `full-node` feature.

File: node/service/src/chain_spec.rs [hunk 1]
```diff
@@ -71,16 +74,24 @@ pub struct Extensions {
 }
 
 /// The `ChainSpec` parameterized for the polkadot runtime.
+#[cfg(feature = "polkadot-native")]
 pub type PolkadotChainSpec = service::GenericChainSpec<polkadot::GenesisConfig, Extensions>;
 
+// Dummy chain spec, in case when we don't have the native runtime.
+pub type DummyChainSpec = service::GenericChainSpec<(), Extensions>;
```
Comment (ordian) [inline node/service/src/chain_spec.rs:81]: not sure if this is correct

**Comment by librelois** [comment]
It would be nice to use `static_assertions` to make sure at compile time that the features combination is valid: https://docs.rs/static_assertions/1.1.0/static_assertions/macro.assert_cfg.html

**Comment by ordian** [comment]
We could also use `std::compile_error` as was suggested here: https://github.com/paritytech/polkadot/pull/3189#discussion_r646712678.

**Comment by ordian** [comment]
@bkchr @tomusdrw please take a look again

**Comment by ordian** [comment]
> Alternatively, could we patch the rococo-runtime crate and use a local version (rialto/millau)?
That way we could have both native runtime, without copying the whole node stuff over.

This PR allows only this approach, but you'd have only one native runtime if you disable polkadot and enable rococo feature.

File: node/service/src/chain_spec.rs (hunk 1 above)
Comment (bkchr) [reply node/service/src/chain_spec.rs:81]: Yeah should be fine.

File: scripts/gitlab/check_no_default_features.sh [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+set -e
+
+pushd node/service && cargo check --no-default-features && popd
+pushd cli && cargo check --no-default-features --features "service" && popd
```
Comment (bkchr) [inline scripts/gitlab/check_no_default_features.sh:6]: Enabling a feature defeats a little bit the name of this file, but fine :P :D

**Review by bkchr — APPROVED** [review]

File: node/service/src/chain_spec.rs [hunk 3]
```diff
@@ -71,16 +74,24 @@ pub struct Extensions {
...
+// Dummy chain spec, but that is fine when we don't have the native runtime.
+#[cfg(not(feature = "polkadot-native"))]
+pub type PolkadotChainSpec = DummyChainSpec;
+
 /// The `ChainSpec` parameterized for the kusama runtime.
 #[cfg(feature = "kusama-native")]
 pub type KusamaChainSpec = service::GenericChainSpec<kusama::GenesisConfig, Extensions>;
 
 /// The `ChainSpec` parameterized for the kusama runtime.
 // This actually uses the polkadot chain spec, but that is fine when we don't have the native runtime.
```
Comment (ordian) [inline node/service/src/chain_spec.rs:92]:
```suggestion
// Dummy chain spec, but that is fine when we don't have the native runtime.
```

File: node/service/src/chain_spec.rs [hunk 4]
```diff
@@ -89,7 +100,7 @@ pub type WestendChainSpec = service::GenericChainSpec<westend::GenesisConfig, Ex
 /// The `ChainSpec` parameterized for the westend runtime.
 // This actually uses the polkadot chain spec, but that is fine when we don't have the native runtime.
```
Comment (ordian) [inline node/service/src/chain_spec.rs:101]:
```suggestion
// Dummy chain spec, but that is fine when we don't have the native runtime.
```

File: node/service/src/chain_spec.rs [hunk 5]
```diff
@@ -98,7 +109,7 @@ pub type RococoChainSpec = service::GenericChainSpec<RococoGenesisExt, Extension
 /// The `ChainSpec` parameterized for the rococo runtime.
 // This actually uses the polkadot chain spec, but that is fine when we don't have the native runtime.
```
Comment (ordian) [inline node/service/src/chain_spec.rs:110]:
```suggestion
// Dummy chain spec, but that is fine when we don't have the native runtime.
```

**Review by svyatonik — APPROVED** [review]

**Comment by ordian** [comment]
bot merge

**Pull request merged**